mixer/cmd/reset.go

```diff
@@ -0,0 +1,271 @@
+By default, the value of PREVIOUS_MIX_VERSION 
+in mixer.state will be used to define the last build. 
+This command can be used to roll back the mixer state 
+in case of a build failure or in case the user 
+wants to roll back to a previous version.`,
+	RunE: runReset,
+}
+
+type resetFlags struct {
+	toVersion int32
+	clean     bool
+}
+
+var resetCmdFlags resetFlags
+
+func init() {
+	RootCmd.AddCommand(resetCmd)
+
+	resetCmd.Flags().Int32Var(&resetCmdFlags.toVersion, "to", -1, "Reset to a specific mix version, default = PREVIOUS_MIX_VERSION")
+	resetCmd.Flags().BoolVar(&resetCmdFlags.clean, "clean", false, "Deletes all files with versions bigger than the one provided")
+}
+
+func runReset(cmd *cobra.Command, args []string) error {
+	if err := checkRoot(); err != nil {
+		fail(err)
+	}
+
+	b, err := builder.NewFromConfig(configFile)
+	if err != nil {
+		return err
+	}
+
+	// if toVersion provided by the user, replace the mixer version
+	if resetCmdFlags.toVersion >= 0 {
+		b.MixVer = strconv.Itoa(int(resetCmdFlags.toVersion))
+		b.MixVerUint32 = uint32(resetCmdFlags.toVersion)
+	} else {
+		// assuming mixer.state file has the correct info
+		b.MixVer = b.State.Mix.PreviousMixVer
+		b.MixVerUint32, err = parseUint32(b.State.Mix.PreviousMixVer)
+		if err != nil {
+			return err
+		}
+		fmt.Println("Reseting to default PREVIOUS_MIX_VERSION ", b.State.Mix.PreviousMixVer)
+	}
+
+	if b.MixVer == "0" {
+		b.State.LoadDefaults(b.Config)
+		b.State.Mix.PreviousMixVer = "0"
+		if err := b.State.Save(); err != nil {
+			fail(err)
+		}
+
+		if err := ioutil.WriteFile(filepath.Join(b.Config.Builder.VersionPath, b.MixVerFile), []byte("10"), 0644); err != nil {
+			return err
+		}
+
+		if !resetCmdFlags.clean {
+			fmt.Println("Reset completed.")
+			return nil
+		}
+		err := os.RemoveAll(b.Config.Builder.ServerStateDir)
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+		fmt.Println("Reset completed.")
+		return nil
+	}
+
+	// find the new previous version by reading the Manifest.Mom file for the new current version
+	momFile := filepath.Join(b.Config.Builder.ServerStateDir, "www", b.MixVer, "Manifest.MoM")
+	mom, err := swupd.ParseManifestFile(momFile)
+	if err != nil {
+		return err
+	}
+	b.State.Mix.PreviousMixVer = strconv.Itoa(int(mom.Header.Previous))
+	currentMixFormatInt := mom.Header.Format
+
+	// Make sure FORMAT in mixer.state has the same value as update/www/<previousMixVer>/format
+	var format string
+	if mom.Header.Previous != 0 {
+		format, err = b.GetFormatForVersion(b.State.Mix.PreviousMixVer)
+	} else {
+		format = strconv.Itoa(int(currentMixFormatInt))
+	}
+	b.State.Mix.Format = strings.TrimSpace(format)
+
+	// Change upstreamVersion file content to the new current version upstreamver
+	var lastStableMixUpstreamVersion string
+	lastStableMixUpstreamVersion, err = b.GetLocalUpstreamVersion(b.MixVer)
+	filename := filepath.Join(b.Config.Builder.ServerStateDir, "www", b.MixVer, "upstreamver")
```

Comment:
This line should be after the error handling of GetLocalUpstreamVersion(). Can cause lint issues.